### PR TITLE
[Scenes] Added the color-control cluster handler for scenes EFS

### PR DESCRIPTION
**Pull request by lpbeliveau-silabs**
Enables color control through scenes actions by adding a default handler to the color control server.

### CODE REVIEW

File: src/app/clusters/color-control-server/color-control-server.cpp [hunk 1]
```diff
@@ -25,11 +25,303 @@
...
+        kEnhancedCurrentHue,
+        kCurrentSaturation,
+        kColorLoopActive,
+        kColorLoopDirection,
+        kColorLoopTime,
+        kColorTemperatureMireds,
+        kEnhancedColorMode,
+    };
+
+    // As per spec, 2 attributes are scenable in the level control cluster
```
Comment (bzbarsky-apple) [inline src/app/clusters/color-control-server/color-control-server.cpp:54]: This comment does not match the code.

File: src/app/clusters/color-control-server/color-control-server.cpp [hunk 2]
```diff
@@ -25,11 +25,303 @@
...
 using namespace chip;
 using namespace chip::app::Clusters;
 using namespace chip::app::Clusters::ColorControl;
 using chip::Protocols::InteractionModel::Status;
 
+#ifdef EMBER_AF_PLUGIN_SCENES
+class DefaultColorControlSceneHandler : public scenes::DefaultSceneHandlerImpl
+{
+public:
+    enum ColorControlEFS : uint8_t
```
Comment (bzbarsky-apple) [inline src/app/clusters/color-control-server/color-control-server.cpp:41]: Should this be an enum class?

File: src/app/clusters/color-control-server/color-control-server.cpp [hunk 3]
```diff
@@ -25,11 +25,303 @@
...
+        kCurrentSaturation,
+        kColorLoopActive,
+        kColorLoopDirection,
+        kColorLoopTime,
+        kColorTemperatureMireds,
+        kEnhancedColorMode,
+    };
+
+    // As per spec, 2 attributes are scenable in the level control cluster
+    static constexpr uint8_t kColorControlMaxScenableAttributes = 9;
```
Comment (bzbarsky-apple) [inline src/app/clusters/color-control-server/color-control-server.cpp:55]: Ideally this would be somehow derived from the enum class, since it needs to stay in sync with it.

Failing that, a comment is needed here and at the end of the enum class about them needing to be in sync.

File: src/app/clusters/color-control-server/color-control-server.cpp [hunk 4]
```diff
@@ -25,11 +25,303 @@
...
+
+    DefaultColorControlSceneHandler() = default;
+    ~DefaultColorControlSceneHandler() override {}
+
+    // Default function for ColorControl cluster, only puts the ColorControl cluster ID in the span if supported on the caller
+    // endpoint
+    void GetSupportedClusters(EndpointId endpoint, Span<ClusterId> & clusterBuffer) override
+    {
+        ClusterId * buffer = clusterBuffer.data();
+        if (emberAfContainsServer(endpoint, ColorControl::Id) && clusterBuffer.size() >= 1)
```
Comment (bzbarsky-apple) [inline src/app/clusters/color-control-server/color-control-server.cpp:65]: If this tests false, shouldn't we reduce the size of the buffer to 0?  Otherwise we are returning whatever size it has incoming, with garbage inside....

File: src/app/clusters/color-control-server/color-control-server.cpp [hunk 5]
```diff
@@ -25,11 +25,303 @@
...
+    {
+        using AttributeValuePair = Scenes::Structs::AttributeValuePair::Type;
+
+        AttributeValuePair pairs[kColorControlMaxScenableAttributes];
+        size_t attributeCount = 0;
+
+        if (ColorControlServer::Instance().HasFeature(endpoint, ColorControlServer::Feature::kXy))
+        {
+            uint16_t xValue = 0x616B; // Default X value according to spec
+            Attributes::CurrentX::Get(endpoint, &xValue);
```
Comment (bzbarsky-apple) [inline src/app/clusters/color-control-server/color-control-server.cpp:93]: This is assuming that on failure the out param will not be written to.  Does the API promise that?

Same in other places here.

File: src/app/clusters/color-control-server/color-control-server.cpp [hunk 6]
```diff
@@ -25,11 +25,303 @@
...
+
+        uint8_t attIdx;
+        while (pair_iterator.Next())
+        {
+            auto & decodePair = pair_iterator.GetValue();
+
+            switch (decodePair.attributeID)
+            {
+            case Attributes::CurrentX::Id:
+                attIdx = ColorControlEFS::kCurrentX;
```
Comment (bzbarsky-apple) [inline src/app/clusters/color-control-server/color-control-server.cpp:175]: I'm a little confused by this part.

So we are reading the id in the entry, mapping that to an index in some array, then putting the value in the entry in the array.

Then walking over the array and applying the values from there.

Why do we want to do that instead of just applying the values as we go.  Is this so that things are atomic in case of TLV decode failure?  Something else?

If we do need to have an intermediate buffer, what I would recommend we do is:

1. Have a constexpr array of AttributeId for the scenable attributes.
2. Declare a buffer with size equal to that array length here.
3. Have a constexpr (so that if you call it with a constant value it's compile-time computed) function for finding the index of an attribute id in the array.
4. Use that to index into the buffer.

That would ensure that if we add a new scenable attribute all of the things here get updated automatically, and we just need to fix the "apply the value" bits below, right?

File: src/app/clusters/color-control-server/color-control-server.cpp [hunk 7]
```diff
@@ -25,11 +25,303 @@
...
+                break;
+            }
+
+            attributeBuffer[attIdx] = decodePair.attributeValue;
+        }
+
+        ReturnErrorOnFailure(pair_iterator.GetStatus());
+
+        // Switch to the mode saved in the scene
+        if (SupportsColorMode(endpoint, static_cast<uint8_t>(attributeBuffer[ColorControlEFS::kEnhancedColorMode])))
```
Comment (bzbarsky-apple) [inline src/app/clusters/color-control-server/color-control-server.cpp:212]: Please store `static_cast<uint8_t>(attributeBuffer[ColorControlEFS::kEnhancedColorMode])` in a temporary, probably one called `colorMode`, so it's not being copy/pasted.

Comment (bzbarsky-apple) [inline src/app/clusters/color-control-server/color-control-server.cpp:212]: Should we be checking for out-of-range values on any of these casts?

File: src/app/clusters/color-control-server/color-control-server.cpp [hunk 8]
```diff
@@ -25,11 +25,303 @@
...
+    /// @return CHIP_NO_ERROR if value as expected, CHIP_ERROR_INVALID_ARGUMENT otherwise
+    CHIP_ERROR ApplyScene(EndpointId endpoint, ClusterId cluster, const ByteSpan & serializedBytes,
+                          scenes::TransitionTimeMs timeMs) override
+    {
+        app::DataModel::DecodableList<Scenes::Structs::AttributeValuePair::DecodableType> attributeValueList;
+
+        ReturnErrorOnFailure(DecodeAttributeValueList(serializedBytes, attributeValueList));
+
+        size_t attributeCount = 0;
+        uint32_t attributeBuffer[kColorControlMaxScenableAttributes];
```
Comment (bzbarsky-apple) [inline src/app/clusters/color-control-server/color-control-server.cpp:160]: So this contains random values, right?  What if some of the values are missing from the attributeValueList?

File: src/app/clusters/color-control-server/color-control-server.cpp [hunk 9]
```diff
@@ -25,11 +25,303 @@
...
+#ifdef EMBER_AF_PLUGIN_COLOR_CONTROL_SERVER_TEMP
+        ColorControlServer::Color16uTransitionState * colorTempTransitionState =
+            ColorControlServer::Instance().getTempTransitionState(endpoint);
+        colorTempTransitionState->finalValue = static_cast<uint16_t>(attributeBuffer[ColorControlEFS::kColorTemperatureMireds]);
+#endif
+
+        // Set Loop Scene Attributes and start loop if scene stored active loop
+        Attributes::ColorLoopDirection::Set(endpoint, static_cast<uint8_t>(attributeBuffer[ColorControlEFS::kColorLoopDirection]));
+        Attributes::ColorLoopTime::Set(endpoint, static_cast<uint8_t>(attributeBuffer[ColorControlEFS::kColorLoopTime]));
+        if (attributeBuffer[ColorControlEFS::kColorLoopActive] == 1)
```
Comment (bzbarsky-apple) [inline src/app/clusters/color-control-server/color-control-server.cpp:249]: Shouldn't this also be conditioned on this endpoint's server supporting the color loop feature?

File: src/app/clusters/color-control-server/color-control-server.cpp [hunk 10]
```diff
@@ -1105,92 +1553,16 @@ bool ColorControlServer::moveToHueAndSaturationCommand(app::CommandHandler * com
...
-    // limit checking:  hue and saturation are 0..254.  Spec dictates we ignore
-    // this and report a constraint error.
-    if ((!isEnhanced && hue > MAX_HUE_VALUE) || saturation > MAX_SATURATION_VALUE)
-    {
-        commandObj->AddStatus(commandPath, Status::ConstraintError);
-        return true;
-    }
-
-    if (!shouldExecuteIfOff(endpoint, optionsMask, optionsOverride))
+    if (!shouldExecuteIfOff(commandPath.mEndpointId, optionsMask, optionsOverride))
```
Comment (bzbarsky-apple) [inline src/app/clusters/color-control-server/color-control-server.cpp:1556]: This is a spec violation.  ConstraintError should be happening if it needs to happen before the shouldExecureIfOff check happens.  Similar in other methods here.

**Review by bzbarsky-apple — CHANGES_REQUESTED** [review]

File: src/app/clusters/color-control-server/color-control-server.cpp (hunk 3 above)
Comment (jmartinez-silabs) [reply src/app/clusters/color-control-server/color-control-server.cpp:55]: You could add the value as the last entry of the enum.

File: src/app/clusters/color-control-server/color-control-server.cpp (hunk 5 above)
Comment (lpbeliveau-silabs) [reply src/app/clusters/color-control-server/color-control-server.cpp:93]: All of the sceneable attributes in here have the same Get accessor part:
`    EmberAfStatus status = emberAfReadAttribute(endpoint, Clusters::ColorControl::Id, Id, readable, sizeof(temp));
    VerifyOrReturnError(EMBER_ZCL_STATUS_SUCCESS == status, status);
    if (!Traits::CanRepresentValue(/* isNullable = */ false, temp))
    {
        return EMBER_ZCL_STATUS_CONSTRAINT_ERROR;
    }
    *value = Traits::StorageToWorking(temp);
    return status;`
    
    In which the value only gets modified if the returned status is success.

Comment (bzbarsky-apple) [reply src/app/clusters/color-control-server/color-control-server.cpp:93]: @lpbeliveau-silabs Yes, but the API does not actually guarantee that.  It could change without anyone realizing this code needs to be updated.

File: src/app/clusters/color-control-server/color-control-server.cpp (hunk 9 above)
Comment (lpbeliveau-silabs) [reply src/app/clusters/color-control-server/color-control-server.cpp:249]: Added check on feature.

File: src/app/clusters/color-control-server/color-control-server.cpp (hunk 8 above)
Comment (lpbeliveau-silabs) [reply src/app/clusters/color-control-server/color-control-server.cpp:160]: Got rid of the buffer and used default values where storage was needed.

File: src/app/clusters/color-control-server/color-control-server.cpp (hunk 10 above)
Comment (lpbeliveau-silabs) [reply src/app/clusters/color-control-server/color-control-server.cpp:1556]: I added the constraint check back before !shouldExecuteIfOff where I removed it.

File: src/app/clusters/color-control-server/color-control-server.cpp (hunk 6 above)
Comment (bzbarsky-apple) [reply src/app/clusters/color-control-server/color-control-server.cpp]: > we do not want to take any action before confirming the colorMode and the color loop active parameters.

OK, makes sense.

File: src/app/clusters/color-control-server/color-control-server.cpp (hunk 5 above)
Comment (lpbeliveau-silabs) [reply src/app/clusters/color-control-server/color-control-server.cpp]: Now writing default value on failure.

File: src/app/clusters/color-control-server/color-control-server.cpp [hunk 11]
```diff
@@ -25,11 +25,348 @@
...
+        }
+
+        uint16_t transitionTime10th = static_cast<uint16_t>(timeMs / 100);
+
+        if (loopActiveValue == 1 && ColorControlServer::Instance().HasFeature(endpoint, ColorControlServer::Feature::kColorLoop))
+        {
+            // Set Loop Scene Attributes and start loop if scene stored active loop
+            Attributes::ColorLoopDirection::Set(endpoint, loopDirectionValue);
+            Attributes::ColorLoopTime::Set(endpoint, loopTimeValue);
+            // Tries to applie color control loop
```
Comment (bzbarsky-apple) [inline src/app/clusters/color-control-server/color-control-server.cpp:297]:
```suggestion
            // Tries to apply color control loop
```

File: src/app/clusters/color-control-server/color-control-server.cpp [hunk 12]
```diff
@@ -850,6 +1197,137 @@ EmberEventControl * ColorControlServer::configureHSVEventControl(EndpointId endp
...
+
+    // Handle color mode transition, if necessary.
+    handleModeSwitch(endpoint, COLOR_MODE_HSV);
+
+    // now, kick off the state machine.
+    initSaturationTransitionState(endpoint, colorSaturationTransitionState);
+    colorSaturationTransitionState->finalValue     = saturation;
+    colorSaturationTransitionState->stepsRemaining = transitionTime;
+    colorSaturationTransitionState->stepsTotal     = transitionTime;
+    colorSaturationTransitionState->endpoint       = endpoint;
```
Comment (bzbarsky-apple) [inline src/app/clusters/color-control-server/color-control-server.cpp:1230]: Where did the setting of `colorSaturationTransitionState->timeRemaining` go?

Please double-check all these refactored functions carefully and document in the PR any changes to them that were not just code movement and the reasons for the changes.

File: src/app/clusters/color-control-server/color-control-server.cpp [hunk 13]
```diff
@@ -850,6 +1197,137 @@ EmberEventControl * ColorControlServer::configureHSVEventControl(EndpointId endp
...
+    colorSaturationTransitionState->lowLimit       = MIN_SATURATION_VALUE;
+    colorSaturationTransitionState->highLimit      = MAX_SATURATION_VALUE;
+
+    SetHSVRemainingTime(endpoint);
+
+    // kick off the state machine:
+    scheduleTimerCallbackMs(configureHSVEventControl(endpoint), TRANSITION_UPDATE_TIME_MS.count());
```
Comment (bzbarsky-apple) [inline src/app/clusters/color-control-server/color-control-server.cpp:1237]: Again, why the change?  This used to schedule for 0 if incoming transitionTime is 0...

File: src/app/clusters/color-control-server/color-control-server.cpp (hunk 12 above)
Comment (lpbeliveau-silabs) [reply src/app/clusters/color-control-server/color-control-server.cpp]: Reverted unintentional changes, only change kept was the order in which `!shouldExecuteIfOff`. This was done so the new "move to" helpers have a coherent behaviour with the already existing moveToColorTemp.

**Review by lpbeliveau-silabs — COMMENTED** [review]
Applied suggestions.

File: src/app/clusters/color-control-server/color-control-server.cpp (hunk 13 above)
Comment (lpbeliveau-silabs) [reply src/app/clusters/color-control-server/color-control-server.cpp]: reverted.

File: src/app/clusters/color-control-server/color-control-server.cpp [hunk 14]
```diff
@@ -1806,7 +2166,23 @@ bool ColorControlServer::moveToColorCommand(app::CommandHandler * commandObj, co
...
+{
+    if (!shouldExecuteIfOff(commandPath.mEndpointId, commandData.optionsMask, commandData.optionsOverride))
+    {
+        commandObj->AddStatus(commandPath, Status::Success);
+        return true;
+    }
+
+    Status status = moveToColor(commandData.colorX, commandData.colorY, commandData.transitionTime, commandPath.mEndpointId);
+#ifdef EMBER_AF_PLUGIN_SCENES
+    //  the scene has been changed (the value of color temp has changed)
```
Comment (bzbarsky-apple) [inline src/app/clusters/color-control-server/color-control-server.cpp:2183]: Again, where does this color temp change happen?

File: src/app/clusters/color-control-server/color-control-server.cpp [hunk 15]
```diff
@@ -2686,6 +3060,11 @@ void emberAfColorControlClusterServerInitCallback(EndpointId endpoint)
 #ifdef EMBER_AF_PLUGIN_COLOR_CONTROL_SERVER_TEMP
     ColorControlServer::Instance().startUpColorTempCommand(endpoint);
 #endif // EMBER_AF_PLUGIN_COLOR_CONTROL_SERVER_TEMP
+#ifdef EMBER_AF_PLUGIN_SCENES
+    // Registers Scene handlers for the level control cluster on the server
```
Comment (bzbarsky-apple) [inline src/app/clusters/color-control-server/color-control-server.cpp:3064]: What does this have to do with level control?

File: src/app/clusters/color-control-server/color-control-server.cpp [hunk 16]
```diff
@@ -1370,62 +1763,31 @@ bool ColorControlServer::moveSaturationCommand(app::CommandHandler * commandObj,
...
-    uint8_t optionsMask     = commandData.optionsMask;
-    uint8_t optionsOverride = commandData.optionsOverride;
-    EndpointId endpoint     = commandPath.mEndpointId;
-    Status status           = Status::Success;
-
-    Color16uTransitionState * colorSaturationTransitionState = getSaturationTransitionState(endpoint);
-    VerifyOrExit(colorSaturationTransitionState != nullptr, status = Status::UnsupportedEndpoint);
-
-    // limit checking:  hue and saturation are 0..254.  Spec dictates we ignore
-    // this and report a malformed packet.
```
[the remark this answers is not in the document]
Comment (lpbeliveau-silabs) [reply src/app/clusters/color-control-server/color-control-server.cpp]: The code seemed explicit in this case and also this seemed like a wrongly copy pasted comment from the move to hue and saturation comment (not that I would ever copy paste comments myself...)
I added it back and removed the hue bits.

File: src/app/clusters/color-control-server/color-control-server.cpp [hunk 17]
```diff
@@ -1126,71 +1574,16 @@ bool ColorControlServer::moveToHueAndSaturationCommand(app::CommandHandler * com
...
-    SetHSVRemainingTime(endpoint);
-
-    // kick off the state machine:
-    scheduleTimerCallbackMs(configureHSVEventControl(endpoint), transitionTime ? TRANSITION_UPDATE_TIME_MS.count() : 0);
-
-exit:
+    Status status = moveToHueAndSaturation(hue, saturation, transitionTime, isEnhanced, commandPath.mEndpointId);
+#ifdef EMBER_AF_PLUGIN_SCENES
+    //  the scene has been changed (the value of color temp has changed)
+    Scenes::ScenesServer::Instance().MakeSceneInvalid(commandPath.mEndpointId);
```
[the remark this answers is not in the document]
Comment (lpbeliveau-silabs) [reply src/app/clusters/color-control-server/color-control-server.cpp]: I just removed all these scene invalid comments, I don't think they add any more to the code.

**Review by bzbarsky-apple — APPROVED** [review]

**Review by jmartinez-silabs — APPROVED** [review]
